packages/langium/src/lsp/call-hierarchy-provider.ts

```diff
@@ -0,0 +1,107 @@
+        this.nameProvider = services.references.NameProvider;
+        this.documents = services.shared.workspace.LangiumDocuments;
+        this.references = services.references.References;
+    }
+
+    prepareCallHierarchy(document: LangiumDocument<AstNode>, params: CallHierarchyPrepareParams): CallHierarchyItem[] | null {
+        const rootNode = document.parseResult.value;
+        const targetNode = findDeclarationNodeAtOffset(rootNode.$cstNode, document.textDocument.offsetAt(params.position), this.grammarConfig.nameRegexp);
+        if (!targetNode) {
+            return null;
+        }
+
+        const declarationNode = this.references.findDeclaration(targetNode);
+        if (!declarationNode) {
+            return null;
+        }
+
+        return this.getCallHierarchyItems(declarationNode?.element, document);
+    }
+
+    protected getCallHierarchyItems(targetNode: AstNode, document: LangiumDocument<AstNode>): CallHierarchyItem[] {
+        const nameNode = this.nameProvider.getNameNode(targetNode);
+        if (!nameNode) {
+            return [];
+        }
+
+        return [{
+            kind: SymbolKind.Method,
+            name: nameNode.text,
+            range: nameNode.range,
+            selectionRange: nameNode.range,
+            uri: document.uri.toString()
+        }];
+    }
+
+    incomingCalls(params: CallHierarchyIncomingCallsParams): CallHierarchyIncomingCall[] | null {
+        const document = this.documents.getOrCreateDocument(URI.parse(params.item.uri));
+        const rootNode = document.parseResult.value;
+        const targetNode = findDeclarationNodeAtOffset(rootNode.$cstNode, document.textDocument.offsetAt(params.item.range.start), this.grammarConfig.nameRegexp);
+        if (!targetNode) {
+            return null;
+        }
+
+        const references = this.references.findReferences(targetNode.element, {includeDeclaration: false, onlyLocal: false}).toArray();
+        return this.getIncomingCalls(references);
+    }
+
+    /**
+     * Override this method to collect the incoming calls for your language
+     */
+    protected getIncomingCalls(_references: ReferenceDescription[]): CallHierarchyIncomingCall[] | null {
+        return null;
+    }
+
+    outgoingCalls(params: CallHierarchyOutgoingCallsParams): CallHierarchyOutgoingCall[] | null {
+        const document = this.documents.getOrCreateDocument(URI.parse(params.item.uri));
+        const rootNode = document.parseResult.value;
+        const targetNode = findDeclarationNodeAtOffset(rootNode.$cstNode, document.textDocument.offsetAt(params.item.range.start), this.grammarConfig.nameRegexp);
+        if (!targetNode) {
+            return null;
+        }
+        return this.getOutgoingCalls(targetNode.element);
+    }
+
+    /**
+     * Override this method to collect the outgoing calls for your language
+     */
+    protected getOutgoingCalls(_node: AstNode): CallHierarchyOutgoingCall[] | null {
```

Comment:
Minor: In line with the previous suggestions, this should be made abstract.
```suggestion
    protected abstract getOutgoingCalls(node: AstNode): CallHierarchyOutgoingCall[] | undefined {
```